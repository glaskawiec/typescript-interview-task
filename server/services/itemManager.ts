import { passwords } from '../data';

let items = [...passwords];

export const updateItem = (updatedItem) => {
  items = items.map((item) => (item.id === updatedItem.id ? updatedItem : item));
};

export const getItems = () => items;
