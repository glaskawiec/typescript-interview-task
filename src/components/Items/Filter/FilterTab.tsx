import { FC } from 'react';
import { useHistory } from 'react-router-dom';

interface IFilterTab {
  title: string;
  count: number;
  path: string;
  testId?: string;
}

const FilterTab: FC<IFilterTab> = ({ title, count, path, testId }) => {
  const { push } = useHistory();

  return (
    <div className='filter-tab' data-testid={testId} onClick={() => push(path)}>
      {`${title} (${count})`}
    </div>
  );
};

export default FilterTab;
