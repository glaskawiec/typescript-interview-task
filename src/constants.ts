export enum Paths {
  Login = '/login',
  Items = '/items',
  Weak = '/items/weak',
  Reused = '/items/reused',
  Old = '/items/old',
  Root = '/',
}

export enum API {
  Login = 'api/login',
  Logout = 'api/logout',
  Items = 'api/items',
  User = 'api/user',
}

export const THIRTY_DAYS_TIMESTAMP = 30 * 24 * 60 * 60 * 1000;
