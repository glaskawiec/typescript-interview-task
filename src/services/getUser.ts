import getUrl from '~/utils/getUrl';
import { API } from '~/constants';

interface UserResponse {
  username: string;
  email: string;
  id: string;
  error?: {
    message: string;
  };
}

const getUser = async (): Promise<UserResponse> => {
  const response = await fetch(getUrl(API.User), {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  const data = (await response.json()) as UserResponse;

  if (data?.error) {
    throw new Error(data.error.message);
  }

  return data;
};

export default getUser;
