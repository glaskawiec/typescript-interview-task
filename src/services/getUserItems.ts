import { API } from '../constants';
import getUrl from '../utils/getUrl';

export interface IItem {
  id: string;
  title: string;
  description: string;
  password: string;
  createdAt: string;
}

interface ItemsResponse {
  items: Array<IItem>;
  error?: {
    message: string;
  };
}

const getUserItems = async (userId?: string): Promise<Array<IItem>> => {
  const url = getUrl(API.Items, {
    userId,
  });

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  const data = (await response.json()) as ItemsResponse;

  if (data?.error) {
    throw new Error(data.error.message);
  }

  return data.items;
};

export default getUserItems;
