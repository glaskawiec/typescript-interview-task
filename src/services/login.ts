import { API } from '~/constants';
import getUrl from '../utils/getUrl';

interface LoginResponse {
  id: string;
  email: string;
  token: string;
  error?: {
    message: string;
  };
}

const login = async (username: string, password: string) => {
  const response = await fetch(getUrl(API.Login), {
    method: 'POST',
    body: JSON.stringify({ username, password }),
    headers: {
      'Content-Type': 'application/json',
    },
  });
  const data = (await response.json()) as LoginResponse;

  if (!response.ok) {
    throw new Error(data.error.message);
  }

  const { token } = data;
  localStorage.setItem('token', token);
};

export default login;
