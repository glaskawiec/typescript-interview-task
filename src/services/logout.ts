import getUrl from '~/utils/getUrl';
import { API } from '~/constants';

const logout = async () => {
  const url = getUrl(API.Logout);

  const response = await fetch(url, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  if (response.ok) {
    localStorage.removeItem('token');
  }
};

export default logout;
