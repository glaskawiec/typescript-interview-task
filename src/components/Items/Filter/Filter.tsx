import { FC } from 'react';
import { Paths } from '../../../constants';
import { IItem } from '../../../services/getUserItems';
import FilterTab from './FilterTab';
import './filter-style.scss';
import itemHasWeakPassword from '../../../utils/itemHasWeakPassword';
import itemHasReusedPassword from '../../..//utils/itemHasReusedPassword';
import itemIsOld from '../../../utils/itemIsOld';

interface IFilter {
  items: Array<IItem>;
}

const Filter: FC<IFilter> = ({ items }) => {
  const weakItemsCount = items.reduce(
    (count, item) => (itemHasWeakPassword(item) ? count + 1 : count),
    0,
  );

  const reusedItemsCount = items.reduce(
    (count, item) => (itemHasReusedPassword(item, items) ? count + 1 : count),
    0,
  );

  const oldItemsCount = items.reduce((count, item) => (itemIsOld(item) ? count + 1 : count), 0);

  return (
    <div className='filter'>
      <FilterTab title='Weak' testId='weak-filter' count={weakItemsCount} path={Paths.Weak} />
      <FilterTab
        title='Reused'
        testId='reused-filter'
        count={reusedItemsCount}
        path={Paths.Reused}
      />
      <FilterTab title='Old' testId='old-filter' count={oldItemsCount} path={Paths.Old} />
    </div>
  );
};

export default Filter;
