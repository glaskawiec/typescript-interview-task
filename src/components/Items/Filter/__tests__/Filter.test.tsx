import { render } from '@testing-library/react';
import { IItem } from '../../../../services/getUserItems';
import renderer from 'react-test-renderer';
import Filter from '../Filter';
import userEvent from '@testing-library/user-event';
import { Paths } from '../../../../constants';

const pushMock = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: pushMock,
  }),
}));

const items: IItem[] = [
  {
    id: '001',
    title: 'Google',
    description: 'My personal account',
    password: 'Password123',
    createdAt: new Date().toISOString(),
  },
  {
    id: '002',
    title: 'Facebook',
    description: 'Facebook account that I manage',
    password: 'SuperDuper5trong!',
    createdAt: new Date(new Date().setMonth(new Date().getMonth() - 2)).toISOString(),
  },
  {
    id: '003',
    title: 'Github',
    description: 'This is where I store my projects',
    password: 'Password123',
    createdAt: new Date().toISOString(),
  },
  {
    id: '004',
    title: 'Reddit',
    description: 'Front page of internet',
    password: '12345678',
    createdAt: new Date().toISOString(),
  },
  {
    id: '005',
    title: 'Discogs',
    description: 'still waiting for new album',
    password: 'Vardenis2000',
    createdAt: new Date().toISOString(),
  },
  {
    id: '006',
    title: 'keybr',
    description: 'Lets practice',
    password: 'Vardenis2000',
    createdAt: new Date().toISOString(),
  },
  {
    id: '007',
    title: 'hackaday',
    description: 'tech news',
    password: 'password@',
    createdAt: new Date().toISOString(),
  },
  {
    id: '008',
    title: 'soundcloud',
    description: 'music',
    password: 'pavardenis321',
    createdAt: new Date().toISOString(),
  },
  {
    id: '009',
    title: 'discord',
    description: 'rumors',
    password: 'discordPassword123.',
    createdAt: new Date().toISOString(),
  },
  {
    id: '010',
    title: 'airdroid',
    description: 'replace android',
    password: 'pass1',
    createdAt: new Date().toISOString(),
  },
];

const setup = () => {
  return render(<Filter items={items} />);
};

beforeEach(() => {
  pushMock.mockClear();
});

test('should render correctly', () => {
  const tree = renderer.create(<Filter items={items} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should redirect to correct path after clicking on "Weak" filter', async () => {
  const { queryByTestId } = setup();

  const weakFilterButton = queryByTestId('weak-filter');
  await userEvent.click(weakFilterButton);

  expect(pushMock.mock.calls[0][0]).toBe(Paths.Weak);
});

test('should redirect to correct path after clicking on "Old" filter', async () => {
  const { queryByTestId } = setup();

  const weakFilterButton = queryByTestId('old-filter');
  await userEvent.click(weakFilterButton);

  expect(pushMock.mock.calls[0][0]).toBe(Paths.Old);
});

test('should redirect to correct path after clicking on "Reused" filter', async () => {
  const { queryByTestId } = setup();

  const weakFilterButton = queryByTestId('reused-filter');
  await userEvent.click(weakFilterButton);

  expect(pushMock.mock.calls[0][0]).toBe(Paths.Reused);
});
