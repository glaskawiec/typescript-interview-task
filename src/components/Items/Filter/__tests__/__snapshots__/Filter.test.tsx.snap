// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="filter"
>
  <div
    className="filter-tab"
    data-testid="weak-filter"
    onClick={[Function]}
  >
    Weak (4)
  </div>
  <div
    className="filter-tab"
    data-testid="reused-filter"
    onClick={[Function]}
  >
    Reused (4)
  </div>
  <div
    className="filter-tab"
    data-testid="old-filter"
    onClick={[Function]}
  >
    Old (1)
  </div>
</div>
`;
